ProcessLib/ThermoHydroMechanics/ThermoHydroMechanicsFEM-impl.h

```diff
@@ -310,20 +352,46 @@ void ThermoHydroMechanicsLocalAssembler<ShapeFunctionDisplacement,
         //
         // temperature equation, temperature part.
         //
-        auto const lambda = porosity * lambda_f + (1 - porosity) * lambda_s;
-        KTT.noalias() += (dNdx_T.transpose() * lambda * dNdx_T +
-                          dNdx_T.transpose() * velocity * N_p * rho_f * C_f) *
-                         w;
-        // coefficient matrix which is used for calculating the residual
-        auto const heat_capacity =
-            porosity * C_f * rho_f + (1 - porosity) * C_s * rho_sr;
-        MTT.noalias() += N_T.transpose() * heat_capacity * N_T * w;
+        const double c_f =
+            liquid_phase
+                .property(
+                    MaterialPropertyLib::PropertyType::specific_heat_capacity)
+                .template value<double>(vars);
+        auto const fluid_thermal_conductivity =
+            liquid_phase
+                .property(
+                    MaterialPropertyLib::PropertyType::thermal_conductivity)
+                .template value<double>(vars);
+        GlobalDimMatrixType effective_thermal_condictivity =
+            MaterialPropertyLib::formEffectiveThermalConductivity<
+                DisplacementDim>(
+                solid_phase
+                    .property(
+                        MaterialPropertyLib::PropertyType::thermal_conductivity)
+                    .value(vars),
+                fluid_thermal_conductivity, porosity);
+
+        KTT.noalias() +=
+            (dNdx_T.transpose() * effective_thermal_condictivity * dNdx_T +
+             dNdx_T.transpose() * velocity * N_p * fluid_density * c_f) *
+            w;
+
+        auto const effective_volumetric_heat_capacity =
+            porosity * fluid_density * c_f +
+            (1.0 - porosity) * solid_density *
+                solid_phase
+                    .property(MaterialPropertyLib::PropertyType::
+                                  specific_heat_capacity)
+                    .template value<double>(vars);
+
+        MTT.noalias() +=
+            N_T.transpose() * effective_volumetric_heat_capacity * N_T * w;
 
         //
         // temperature equation, pressure part
         //
-        KTp.noalias() += K_over_mu * rho_f * C_f * N_T.transpose() *
-                         (dNdx_T * T).transpose() * dNdx_T * w;
+        KTp.noalias() += fluid_density * c_f * N_T.transpose() *
```

Comment:
Just out of curiosity: Why did you reorder the calculation?

Reply:
Because K_over_mu is changed to matrix now.